### PR TITLE
Add validations and unit tests

### PR DESCRIPTION
**Pull request by balatstar**
The following were implemented in this branch:

Add validations for the User model:
- name must not be blank
- posts_counter must be an integer greater than or equal to zero

Add validations for the Post model:
- title must not be blank and not exceed 250 characters
- comments_counter must be an integer greater than or equal to zero
- likes_counter must be an integer greater than or equal to zero

Add unit tests for all models' methods and validations.

### CODE REVIEW

File: spec/models/post_spec.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    user = User.create(name: 'Roy Batty', posts_counter: 5)
+    post = Post.new(title: 'A' * 251, author: user)
+    expect(post).not_to be_valid
+  end
+
+  it "increments author's posts_counter after create" do
+    user = User.create(name: 'Roy Batty', posts_counter: 5)
+    Post.create(title: 'Sample Title', author: user)
+    expect(user.reload.posts_counter).to eq(6)
+  end
```
Comment (mckent05) [inline spec/models/post_spec.rb]: - Good job done so far, but according to the project requirement, you are to test all customer methods in your model. Please write tests for the `increment_post_counter` and `decrement_post_counter`. Please fix this 🙏 
- Also, please write tests for the increment and decrement posts_comment_counter in the comment model, please fix this to ensure the methods in your models are tested according to the project requirements 😄 
- Also, please write tests for the increment and decrement likes_counter in the likes model, please fix this to ensure the methods in your models are tested according to the project requirements 😄

**Review by mckent05 — CHANGES_REQUESTED** [review]
Hi @balatstar, 

# STATUS: CHANGES REQUIRED ♻️ ♻️ 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
almost there

## Highlights
- All linter checks are passing ✔️ ✔️ 
- Your PR is professional 👍 👍 
- You have done well writing validations and specs for your models 👏 👏 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by balatstar** [comment]
For the next code reviewer: Hi! The previous reviewer asked me to add tests for increment and decrement posts counters but I already had it there (at the point where he added the comment). Please let me know if I'm not understanding correctly.

I've also added tests for incrementing and decrementing likes and comments, but if I misunderstood the one for posts they might be wrong, too. Thanks!

**Review by DammyShittu — APPROVED** [review]
Hi @balatstar,

## APPROVED 🟢

## Highlights
✔️ Passing tests

Great implementation of all the requirements in this milestone. 👏 

_You have done an amazing job 👍_

Your project is complete!  There is nothing else to say other than... it's time to merge it. :shipit: 

Congratulations! 🎉 

## CHEERS AND HAPPY CODING!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by mckent05** [comment]
> For the next code reviewer: Hi! The previous reviewer asked me to add tests for increment and decrement posts counters but I already had it there (at the point where he added the comment). Please let me know if I'm not understanding correctly.
> 
> I've also added tests for incrementing and decrementing likes and comments, but if I misunderstood the one for posts they might be wrong, too. Thanks!

Hello @balatstar, Great work on your project so far. I saw that you tested to check if your `post_counter` increments or decrements upon creating or destroying a post respectively. But you did not exactly use the methods responsible for these actions upon creating and destroying. Instead of using the `reload` method after creating a post, I'd have expected that you called the function responsible for incrementing your post_counter, that way you can be sure the function works as expected. I hope I am clearer?

**Comment by balatstar** [comment]
Hi @mckent05! Thanks for taking the time to reply! I already merged after the 2nd code reviewer's approval, but I appreciate your insight. Just a question, though... since I'm already automatically incrementing and decrementing post_counter whenever a post is created/destroyed, if I call the increment/decrement function again (instead of reload), won't this increment/decrement twice for each post created/destroyed (one automatically and the other one when I call the method again in the test)? Thanks!!

**Comment by mckent05** [comment]
> Hi @mckent05! Thanks for taking the time to reply! I already merged after the 2nd code reviewer's approval, but I appreciate your insight. Just a question, though... since I'm already automatically incrementing and decrementing post_counter whenever a post is created/destroyed, if I call the increment/decrement function again (instead of reload), won't this increment/decrement twice for each post created/destroyed (one automatically and the other one when I call the method again in the test)? Thanks!!

Hello @balatstar, sorry for the late response.Good question, In your models, the increment and decrement method are only called when the data is actually saved to the database, hence why we use the `after_save` module. But when in the test environment, the data is not saved into the development database after it is created so the increment and decrement functions would not be triggered automatically. So when testing the functions would only be triggered when you call them. I hope this answers your question? :smile:

**Pull request merged**